### PR TITLE
EIP-1559 and Rinkeby Testnet support in Swaps

### PR DESCRIPTION
**Pull request by dan437**
## Explanation
This PR:
- Adds support for EIP-1559 in Swaps
- Uses new components for editing gas in Swaps
- Adds support for a new network in Swaps (Rinkeby), because it already has EIP-1559 support

## Manual testing steps
EIP-1559 happy path:
- Switch to the `Rinkeby` Test Network
- Get test ETH
- Click on "Swap" on the main page
- Make a swap
- Open the transaction in a block explorer (e.g. https://rinkeby.etherscan.io/) and check that Txn Type is: `2 (EIP-1559)` 

Non EIP-1559 happy path:
- Use e.g. the Ethereum Mainnet Network
- Click on "Swap" on the main page
- Make a swap
- Open the transaction in a block explorer (e.g. https://etherscan.io/) and check that Txn Type is not: `2 (EIP-1559)` 

## Screenshots
`View Quote` page on Rinkeby, which has EIP-1559 support:
image

Success page for an EIP-1559 transaction:
image

`Customize Gas` modal window for EIP-1559, Basic Options:
image

`Customize Gas` modal window for EIP-1559, Advanced Options:
image

EIP-1559 Txn Type on rinkeby.etherscan.io:
image

EIP-1559 transaction in the `Activity` tab on the main page:
image

`View Quote` on Ethereum testnet, without EIP-1559 support:
image

Success page for a non-EIP 1559 transaction:
image

Non-EIP-1559 transaction in the `Activity` tab on the main page:
image

### CODE REVIEW

File: ui/store/actions.js [hunk 1]
```diff
@@ -2199,6 +2199,22 @@ export function setSwapsTxGasLimit(gasLimit) {
...
 }
 
+export function updateCustomSwapsEIP1559GasParams({
+  gasLimit,
+  maxFeePerGas,
+  maxPriorityFeePerGas,
+}) {
+  return async (dispatch) => {
+    // TODO:Why do we pass true as the 2nd param? It seems the fn only supports gasLimit.
+    await promisifiedBackground.setSwapsTxGasLimit(gasLimit, true);
```
Comment (dan437) [inline ui/store/actions.js:2209]: I can change this to: `setSwapsTxGasLimit(gasLimit)` everywhere

File: app/scripts/controllers/swaps.js [hunk 2]
```diff
@@ -514,7 +535,38 @@ export default class SwapsController {
...
+      } = gasFeeEstimates;
+
+      usedGasPrice = addCurrencies(
+        customMaxPriorityFeePerGas || // Is already in hex WEI.
+          decGWEIToHexWEI(suggestedMaxPriorityFeePerGas),
+        decGWEIToHexWEI(estimatedBaseFee),
+        {
+          aBase: 16,
+          bBase: 16,
+          toNumericBase: 'hex',
```
Comment (dan437) [inline app/scripts/controllers/swaps.js:558]: Just to double check - are these params correct?

File: ui/components/app/transaction-detail-item/transaction-detail-item.component.js [hunk 3]
```diff
@@ -22,14 +22,14 @@ export default function TransactionDetailItem({
         <Typography
           color={detailTitleColor}
           fontWeight={FONT_WEIGHT.BOLD}
-          variant={TYPOGRAPHY.H6}
+          variant={TYPOGRAPHY.H7}
```
Comment (dan437) [inline ui/components/app/transaction-detail-item/transaction-detail-item.component.js:25]: I've changed this because according to our design we should use H7. Benefit is that now text fits on one line: 
image

File: ui/pages/swaps/index.js [hunk 4]
```diff
@@ -111,8 +112,11 @@ export default function Swap() {
   const chainId = useSelector(getCurrentChainId);
   const isSwapsChain = useSelector(getIsSwapsChain);
   const useNewSwapsApi = useSelector(getUseNewSwapsApi);
+  const EIP1559NetworkEnabled = useSelector(isEIP1559Network);
   const fromToken = useSelector(getFromToken);
 
+  useGasFeeEstimates(); // This will pre-load gas fees before going to the View Quote page.
```
Comment (dan437) [inline ui/pages/swaps/index.js:118]: Double check this.

File: ui/pages/swaps/swaps.util.js [hunk 5]
```diff
@@ -93,6 +96,9 @@ const getBaseApi = function (
       return `${baseUrl}/aggregatorMetadata`;
     case 'gasPrices':
       return `${baseUrl}/gasPrices`;
+    case 'suggestedGasFees':
+      // We will probably use the hook "useGasFeeEstimates" instead of this.
+      return `${GAS_API_BASE_URL}/networks/${chainIdDecimal}/suggestedGasFees`;
```
Comment (dan437) [inline ui/pages/swaps/swaps.util.js:101]: I can probably remove this since we use `useGasFeeEstimates` in Swaps, which calls this endpoint.

File: ui/pages/swaps/view-quote/view-quote.js [hunk 6]
```diff
@@ -134,6 +150,8 @@ export default function ViewQuote() {
   const chainId = useSelector(getCurrentChainId);
   const nativeCurrencySymbol = useSelector(getNativeCurrency);
 
+  const gasFeeInputs = useGasFeeInputs('high');
```
Comment (dan437) [inline ui/pages/swaps/view-quote/view-quote.js:153]: For Swaps we want to get 'high' estimations by default.

File: app/scripts/controllers/swaps.js [hunk 7]
```diff
@@ -514,7 +535,38 @@ export default class SwapsController {
...
+          numberOfDecimals: 6,
+        },
+      );
+    } else if (gasEstimateType === GAS_ESTIMATE_TYPES.LEGACY) {
+      usedGasPrice = customGasPrice || decGWEIToHexWEI(gasFeeEstimates.high);
+    } else if (gasEstimateType === GAS_ESTIMATE_TYPES.ETH_GASPRICE) {
+      usedGasPrice =
+        customGasPrice || decGWEIToHexWEI(gasFeeEstimates.gasPrice);
+    } else {
+      usedGasPrice = '0x0';
```
Comment (darkwing) [inline app/scripts/controllers/swaps.js:568]: If you initialize `usedGasPrice` to `0x0`, you can avoid the `else` block; `0x0` would act as the default value.

File: ui/pages/swaps/view-quote/view-quote.js (hunk 6 above)
Comment (darkwing) [reply ui/pages/swaps/view-quote/view-quote.js:153]: Yep!

**Comment by darkwing** [comment]
One thing that I'm seeing that looks weird is that the timing function shows "60" seconds for high, and 45 for medium and low oO.  Weird.

https://user-images.githubusercontent.com/46655/127207237-aca79b1c-474d-4fae-95d5-2a644cf84536.mp4

**Review by hendis85 — APPROVED** [review]

**Comment by dan437** [comment]
> One thing that I'm seeing that looks weird is that the timing function shows "60" seconds for high, and 45 for medium and low oO. Weird.
> WeirdTiming.mp4

It seems that for Low and Medium it says: `Likely in < 45 seconds` and for High: `Very likely in < 60 seconds`. Agree that it's a bit confusing for users and so we might want to update the `<GasTiming>` component in another PR.

File: ui/store/actions.js [hunk 8]
```diff
@@ -2199,6 +2199,21 @@ export function setSwapsTxGasLimit(gasLimit) {
...
+  gasLimit,
+  maxFeePerGas,
+  maxPriorityFeePerGas,
+}) {
+  return async (dispatch) => {
+    await promisifiedBackground.setSwapsTxGasLimit(gasLimit);
+    await promisifiedBackground.setSwapsTxMaxFeePerGas(maxFeePerGas);
+    await promisifiedBackground.setSwapsTxMaxFeePriorityPerGas(
+      maxPriorityFeePerGas,
+    );
```
Comment (darkwing) [inline ui/store/actions.js:2212]: Doesn't need to block merge but can this be a `Promise.all`, just so we aren't successively awaiting?

**Comment by darkwing** [comment]
This is awesome!  
Successful

File: ui/store/actions.js (hunk 8 above)
Comment (dan437) [reply ui/store/actions.js]: That sounds promising!

File: ui/components/app/transaction-detail-item/transaction-detail-item.component.js [hunk 9]
```diff
@@ -39,20 +39,24 @@ export default function TransactionDetailItem({
...
       </div>
       <div className="transaction-detail-item__row">
-        <Typography
-          variant={TYPOGRAPHY.H7}
-          className="transaction-detail-item__subtitle"
-          color={COLORS.UI4}
-        >
-          {subTitle}
-        </Typography>
+        {React.isValidElement(subTitle) ? (
```
Comment (darkwing) [inline ui/components/app/transaction-detail-item/transaction-detail-item.component.js:49]: I'm very much looking forward to this change! :)

File: app/scripts/controllers/transactions/index.js [hunk 10]
```diff
@@ -1200,10 +1200,12 @@ export default class TransactionController extends EventEmitter {
           txMeta.chainId,
         );
 
-        const quoteVsExecutionRatio = `${new BigNumber(tokensReceived, 10)
-          .div(txMeta.swapMetaData.token_to_amount, 10)
-          .times(100)
-          .round(2)}%`;
```
Comment (danjm) [inline app/scripts/controllers/transactions/index.js:1206]: Why is this change needed?

File: shared/constants/swaps.js [hunk 11]
```diff
@@ -54,6 +55,14 @@ export const TEST_ETH_SWAPS_TOKEN_OBJECT = {
   iconUrl: TEST_ETH_TOKEN_IMAGE_URL,
 };
 
+export const RINKEBY_SWAPS_TOKEN_OBJECT = {
```
Comment (danjm) [inline shared/constants/swaps.js:58]: I guess we should include this for QA purposes, but we should remove it before shipping to prod, because swaps does not actually work on rinkeby

File: ui/ducks/swaps/swaps.js [hunk 12]
```diff
@@ -682,7 +715,10 @@ export const signAndSendTransactions = (history, metaMetricsEvent) => {
       .plus(usedQuote.approvalNeeded?.gas || '0x0', 16)
       .toString(16);
     const gasEstimateTotalInUSD = getValueFromWeiHex({
-      value: calcGasTotal(totalGasLimitEstimate, usedGasPrice),
+      value: calcGasTotal(
+        totalGasLimitEstimate,
+        EIP1559Network ? maxFeePerGas : usedGasPrice,
```
Comment (danjm) [inline ui/ducks/swaps/swaps.js:720]: I think this should be `baseFeePerGas + maxPriorityFeePerGas`

Comment (danjm) [reply ui/ducks/swaps/swaps.js:720]: `usedGasPrice` is use to calculate the "estimate", which is what `baseFeePerGas + maxPriorityFeePerGas` would be used for

File: ui/pages/swaps/view-quote/view-quote.js [hunk 13]
```diff
@@ -607,6 +665,24 @@ export default function ViewQuote() {
...
           />
         )}
+
+        {showEditGasPopover && EIP1559Network && (
+          <EditGasPopover
+            transaction={{
+              txParams: {
+                maxFeePerGas,
+                maxPriorityFeePerGas,
+                gasLimit: maxGasLimit,
```
Comment (danjm) [inline ui/pages/swaps/view-quote/view-quote.js:675]: Let's change this to `gas: maxGasLimit`

File: ui/hooks/useGasFeeInputs.js [hunk 14]
```diff
@@ -200,8 +212,8 @@ export function useGasFeeInputs(defaultEstimateToUse = 'medium', transaction) {
       : null,
   );
   const [gasLimit, setGasLimit] = useState(
-    transaction?.txParams?.gas
-      ? Number(hexToDecimal(transaction.txParams.gas))
```
Comment (danjm) [inline ui/hooks/useGasFeeInputs.js:204]: Hmm... this won't work for standard transactions in metamask. `gasLimit` is normally called `gas`

Instead could you call the param passed in from "View-quote" `gas`? (Then I think this change won't be necessary

Comment (danjm) [reply ui/hooks/useGasFeeInputs.js:204]: Also, we need to keep that `Number` conversion present to handle transactions in the standard metamask send flow

Could you convert the value to hex in `View-quote`? You can use a function called `hexToDecimal` for that

Comment (danjm) [reply ui/hooks/useGasFeeInputs.js:204]: I realize I might have been the original author of these changes :sweat_smile:

File: ui/pages/swaps/index.js [hunk 15]
```diff
@@ -111,8 +112,15 @@ export default function Swap() {
   const chainId = useSelector(getCurrentChainId);
   const isSwapsChain = useSelector(getIsSwapsChain);
   const useNewSwapsApi = useSelector(getUseNewSwapsApi);
+  const EIP1559Network = useSelector(isEIP1559Network);
   const fromToken = useSelector(getFromToken);
 
+  if (EIP1559Network) {
```
Comment (danjm) [inline ui/pages/swaps/index.js:118]: Do we have to have this `if` block and eslint disable? Will everything still work if it is removed?

**Review by danjm — CHANGES_REQUESTED** [review]
Identified a few issues that need to be addressed

File: shared/constants/swaps.js (hunk 11 above)
Comment (dan437) [reply shared/constants/swaps.js:58]: What if we just disable swapping for Rinkeby after we are done with QA, at least until our EIP 1559 changes are in production and working? It could be done by removing this line: 
https://github.com/MetaMask/metamask-extension/pull/11635/files#diff-9f9eff0a280f38efd9f071f24d503f51a1f7617a30f67e7e654e157dfc060902R98

Once we verify that EIP 1559 changes work well in production, I can send a quick PR for removing Rinkeby from Swaps.

File: app/scripts/controllers/transactions/index.js (hunk 10 above)
Comment (dan437) [reply app/scripts/controllers/transactions/index.js:1206]: Great question! For some reason I was getting empty `txMeta.txReceipt.logs` on Rinkeby: 
image

When that happens, the `getSwapsTokensReceivedFromTxMeta` fn call returns null (https://github.com/MetaMask/metamask-extension/blob/develop/ui/pages/swaps/swaps.util.js#L707). Any idea why there are on logs on Rinkeby?

File: ui/hooks/useGasFeeInputs.js (hunk 14 above)
Comment (dan437) [reply ui/hooks/useGasFeeInputs.js:204]: Will do the suggested changes.

File: ui/pages/swaps/index.js (hunk 15 above)
Comment (dan437) [reply ui/pages/swaps/index.js:118]: Originally I didn't have this `if` block, but then I realized we don't need to make unnecessary API calls for non EIP-1559 networks in Swaps via this hook. For non EIP-1559 networks we get gas prices like this: https://github.com/MetaMask/metamask-extension/pull/11635/files#diff-f89a0460c9cb04858f4de957f65d6aab869c1d766f11781b1fe8a61f3321620eR199

File: ui/hooks/useGasFeeInputs.js (hunk 14 above)
Comment (dan437) [reply ui/hooks/useGasFeeInputs.js:204]: Done. Btw, on the View Quote page `maxGasLimit` is already in hex, so I can just do this:
image

Edit Gas modal after these changes:
image

File: shared/constants/swaps.js (hunk 11 above)
Comment (danjm) [reply shared/constants/swaps.js:58]: yeah, that makes sense

**Review by danjm — APPROVED** [review]

**Review by darkwing — APPROVED** [review]
This looks excellent!  The *only* thing that I noticed is that when I click "Edit", the "high" radio isn't selected.  I don't want to block on that though -- functionally working great!

**Comment by dan437** [comment]
> This looks excellent! The _only_ thing that I noticed is that when I click "Edit", the "high" radio isn't selected. I don't want to block on that though -- functionally working great!

Thank you! It seems to be working now, I can see the "High" radio button selected after clicking on "Edit":
image

**Comment by PeterYinusa** [comment]
LGTM!
